cloud/linode/firewall/firewalls.go

```diff
@@ -0,0 +1,463 @@
+func (l *LinodeClient) UpdateNodeBalancerFirewall(
+	ctx context.Context,
+	loadBalancerName string,
+	loadBalancerTags []string,
+	service *v1.Service,
+	nb *linodego.NodeBalancer,
+) error {
+	// get the new firewall id from the annotation (if any).
+	_, fwIDExists := service.GetAnnotations()[annotations.AnnLinodeCloudFirewallID]
+	if fwIDExists { // If an ID exists, we ignore everything else and handle just that
+		return l.updateServiceFirewall(ctx, service, nb)
+	}
+
+	// See if a acl exists
+	_, fwACLExists := service.GetAnnotations()[annotations.AnnLinodeCloudFirewallACL]
+	if fwACLExists { // if an ACL exists, but no ID, just update the ACL on the fw.
+		return l.updateNodeBalancerFirewallWithACL(ctx, loadBalancerName, loadBalancerTags, service, nb)
+	}
+
+	// No firewall ID or ACL annotation, see if there are firewalls attached to our nb
+	firewalls, err := l.Client.ListNodeBalancerFirewalls(ctx, nb.ID, &linodego.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	if len(firewalls) == 0 {
+		return nil
+	}
+	if len(firewalls) > 1 {
+		klog.Errorf("Found more than one firewall attached to nodebalancer: %d, firewall IDs: %v", nb.ID, firewalls)
+		return ErrTooManyNBFirewalls
+	}
+
+	err = l.Client.DeleteFirewallDevice(ctx, firewalls[0].ID, nb.ID)
+	if err != nil {
+		return err
+	}
+	// once we delete the device, we should see if there's anything attached to that firewall
+	devices, err := l.Client.ListFirewallDevices(ctx, firewalls[0].ID, &linodego.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	if len(devices) == 0 {
+		// nothing attached to it, clean it up
+		return l.Client.DeleteFirewall(ctx, firewalls[0].ID)
+	}
+	// else let that firewall linger, don't mess with it.
+
+	return nil
+}
+
+// getNodeBalancerDeviceID gets the deviceID of the nodeBalancer that is attached to the firewall.
+func (l *LinodeClient) getNodeBalancerDeviceID(ctx context.Context, firewallID, nbID int) (int, bool, error) {
+	devices, err := l.Client.ListFirewallDevices(ctx, firewallID, &linodego.ListOptions{})
+	if err != nil {
+		return 0, false, err
+	}
+
+	if len(devices) == 0 {
+		return 0, false, nil
+	}
+
+	for _, device := range devices {
+		if device.Entity.ID == nbID {
+			return device.ID, true, nil
+		}
+	}
+
+	return 0, false, nil
+}
+
+// Updates a service that has a firewallID annotation set.
+// If an annotation is set, and the nodebalancer has a firewall that matches the ID, nothing to do
+// If there's more than one firewall attached to the node-balancer, an error is returned as its not a supported use case.
+// If there's only one firewall attached and it doesn't match what's in the annotation, the new firewall is attached and the old one removed
+func (l *LinodeClient) updateServiceFirewall(ctx context.Context, service *v1.Service, nb *linodego.NodeBalancer) error {
+	var newFirewallID int
+	var err error
+
+	fwID := service.GetAnnotations()[annotations.AnnLinodeCloudFirewallID]
+	newFirewallID, err = strconv.Atoi(fwID)
```

Comment:
nit: There is no need to declare `newFirewallID` and `err` this high up.

```suggestion
	fwID := service.GetAnnotations()[annotations.AnnLinodeCloudFirewallID]
	newFirewallID, err := strconv.Atoi(fwID)
```